### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by badarot** [scikit-hep/uproot4#106]
It seems, at least for me, that uproot4 is using twice as memory to load a branch compared to uproot3.

A created a simple but large enough TTree to test this with this script

```c++
#include <array>

void test() {
    TFile file("test.root", "recreate");
    TTree tree("tree", "test memory leak");

    std::array<Float_t, 1000> array;
    tree.Branch("array", &array);

    for (int i = 0; i < 100000; i++) {
        for (auto &x : array)
            x = i;
        tree.Fill();
    }
    file.Write();
    file.Close();
}
```
And them I loaded this file using uproot3 and 4
```python
import uproot
with uproot.open('test.root') as file:
    data = file['tree/array'].array()
```

```python
import uproot4
with uproot4.open('test.root') as file:
    data = file['tree/array'].array(library='np')
```
Checking the memory consumption of those processes I got this 
```bash
$ ps -e -o command,pid,%mem | grep python
python  21032  10.5   # uproot4
python  21256   5.2   # uproot3
```
The uproot4 version is using twice as memory, so it seams the data is duplicated in memory. Next I deleted the data array with `del data` in both processes and that release the memory used by the array but the "duplicated" data remained in uproot4.

```bash
$ ps -e -o command,pid,%mem | grep python
python  21032  5.7  # uproot4
python  21256  0.4  # uproot3
```
For this test I used the 0.23 version of uproot4. I'm doing something wrong or this is indeed a bug?

**Comment by jpivarski** [comment]
What happens if you

```python
import gc
gc.collect()
```

before or after the explicit `data` deletion? Any Python objects that are created will remain in the process's memory even if their reference counts are zero, until collected by the garbage collector. Fortunately, Python's `gc.collect()` does what you tell it to, allowing you to debug these things by explicitly invoking it. (Trying to debug Java's garbage collector is much harder because its equivalent command is a mere suggestion.)

One thing that Uproot 4 does differently than Uproot 3 is:

   * Uproot 3 allocates a destination array and all interpreted TBaskets write into that array as they're read
   * Uproot 4 collects a list of interpreted TBasket arrays and concatenates them when done.

The latter will use twice as much working memory, but it's less prone to error and more flexible (we can do more interpretation in parallel without knowing how long each variable-length structure will be before combining them; Uproot 3 had to delay interpretation of complex structures as [ObjectArrays](https://github.com/scikit-hep/awkward-array/blob/15a63d57659ef9e34baaf7654e54cd3fe5d0e5fc/awkward/array/objects.py#L33), which [notoriously limited functionality](https://github.com/scikit-hep/uproot/issues?q=is%3Aissue+objectarray)).

I've just confirmed that these TBasket arrays are _not_ cached. They should go out of scope when the array is fully constructed. Specifically, the `basket_arrays` is a dict from basket number to the temporary basket array, and the `basket_arrays` is a value of the `branchid_arrays` dict, which maps from a TBranch identifier to the `basket_arrays` for that TBranch (for functions that fill multiple branches in one call). That `branchid_arrays` dict goes out of scope when this function finishes:

https://github.com/scikit-hep/uproot4/blob/28a867d4b2215069c6eeebebc46132ed11f8f8e7/uproot4/behaviors/TBranch.py#L3150-L3273

That's the only function that fills arrays; [TBranch.array](https://uproot4.readthedocs.io/en/latest/uproot4.behaviors.TBranch.TBranch.html#array), [HasBranches.arrays](https://uproot4.readthedocs.io/en/latest/uproot4.behaviors.TBranch.HasBranches.html), [uproot4.iterate](https://uproot4.readthedocs.io/en/latest/uproot4.behaviors.TBranch.iterate.html), [uproot4.concatenate](https://uproot4.readthedocs.io/en/latest/uproot4.behaviors.TBranch.concatenate.html), and [uproot4.lazy](https://uproot4.readthedocs.io/en/latest/uproot4.behaviors.TBranch.lazy.html) all call into it.

So even though we (intentionally) use twice as much working memory now, those objects _should be_ out of scope and deleted on the next garbage collector pass, assuming that the garbage collector actually runs.

**Comment by badarot** [comment]
Hi Jim, thank you for the really fast and complete answer, I'm new to uproot I am still reading and understanding the code that you pointed.

In the mean time I tried to use the garbage collector as you asked and didn't see any change. 

```python
>>> import uproot4
>>> with uproot4.open('test.root') as file:
...     data = file['tree/array'].array(library='np')
...  
>>> import gc    # first ps
>>> gc.collect() # second ps
0
>>> del data     # third ps
>>> gc.collect() # forth ps
0
```
```bash
❯ ps -e -o command,pid,%mem | grep python  # after read the file
python                        57623 10.6
❯ ps -e -o command,pid,%mem | grep python  # after the first gc.collect() call
python                        57623 10.6
❯ ps -e -o command,pid,%mem | grep python # after the del data
python                        57623  5.7
❯ ps -e -o command,pid,%mem | grep python # after the second gc.collect() call
python                        57623  5.7
```
So even after the deletion of the array and the garbage collection call the process seems to hold the TTree data in memory. That 5.7 % of memory is roughly the size of the TTree.

**Comment by henryiii** [comment]
PS: The garbage collector in CPython is usually set to run every line.

**Comment by tamasgal** [comment]
As @henryiii said, the GC might mess up the memory measurements.

Alternatively to the manual collection of `gc.collect()` you can set the garbage collector in "aggressive mode":

```python
import gc
gc.set_threshold((1, 1, 1))
```

or artificially decrease the available memory to see if the memory is still required.

**Comment by jpivarski** [comment]
CPython doesn't collect all garbage on every line, and I was only suggesting `gc.collect()` as a debugging tool, not for production code.

I'll look into this, to see if the memory is really leaking or just isn't being reclaimed.

By the way, although the basket arrays are not put in the cache, the final array is. Are you accounting for that?

https://uproot4.readthedocs.io/en/latest/basic.html#caching-and-memory-management

**Comment by badarot** [comment]
> By the way, although the basket arrays are not put in the cache, the final array is. Are you accounting for that?

No, I wasnt aware of that. So the line `data = file['tree/array'].array(library='np')` fill the cache with the branch data and pass a reference of it to the `data` variable or I get a full copy of the cache, ending up with the two copies in memory?

**Comment by jpivarski** [comment]
Man, I keep meaning to get to this and I can't. Maybe an easy way to turn the cache off is

```python
uproot4.object_cache = None
uproot4.array_cache = None
```

at the beginning of the program.

**Comment by jpivarski** [comment]
One of the things that came out of this was that it forced me to think hard about the caching policy. Sharing a cache among files is not a good idea because you, the user, would expect that deleting the arrays you get from a file and deleting the file object would be sufficient to remove all trace of those arrays from memory. Not so if there's a shared cache, so I'm getting rid of that. Files still have individual caches for the reasons described in the above link, but when you get rid of a file, everything that had been stuck to it will now be gone.

The actual memory leak wasn't related to the caches, though. It was in the ThreadPool and ResourceThreadPool executors, which tie threads to the file handles they manage so that everything can be eliminated when a file's `close` or `__exit__` is called (leaves a `with` block).

Here it is:

```diff
    def run(self):
        """
        Listens to the :py:attr:`~uproot4.source.futures.ResourceWorker.work_queue`
        and executes each :py:class:`~uproot4.source.futures.ResourceFuture` it
        receives (with :py:attr:`~uproot4.source.futures.ResourceWorker.resource` as
        its first argument) until it receives None.
        """
+       future = None
        while True:
+           del future   # don't hang onto a reference while waiting for more work
            future = self._work_queue.get()
            if future is None:
                break
            assert isinstance(future, ResourceFuture)
            future._run(self._resource)
```

Without the new lines, the previous `future` is still in scope while the `self._work_queue.get()` waits for new work because the assignment can't happen until the function on the right-hand side finishes. The `future` contains a basket array from the previous task, and if you have enough threads, those baskets add up. They will be deleted as soon as new work comes in, so this is not a permanent memory leak (the memory usage would not grow without bound), but it is the cause of your surprising behavior.

With the fix, here's what happens:

```python
>>> import uproot4, gc
>>> branch = uproot4.open("test.root:tree/array", array_cache=None)
>>> gc.collect()
0
>>> # RSS is at 35812
>>> array = branch.array(library="np")
>>> # RSS is at 853356
>>> gc.collect()
57144
>>> # RSS is at 452968
>>> branch.close()
>>> del branch
>>> gc.collect()
147
>>> # RSS is at 448472
>>> del array
>>> gc.collect()
0
>>> # RSS is at 57588
```

Immediately after we read the array, there's 798 MiB of temporary data whose reference count is zero but the garbage collector hasn't gotten rid of it. I assume that's because of cyclic references, but I haven't investigated. Calling `gc.collect()` explicitly finds those objects and eliminates them, but in a memory-limited situation, the garbage collector would have been triggered automatically.

Closing the file doesn't release much, 4.3 MiB of various things related to those threads and file handles; maybe also the memory-mapped file. (I don't know how that gets counted in the RSS.)

But when we delete the array itself (which goes away immediately; no cyclic references), the 381 MiB array goes away, and we're left with a baseline that's 21 MiB higher than when we started. It may be that some libraries were imported in the process. On another run, this was 11 MiB. If I step through this procedure several times in the same process, we come back to this new baseline (and bounce around by ±10 MiB). Since it bounces around, this is in the noise.
